[#2016] Remove hash symbol from URL when decoding hash

jonasongg opened the pull request:
<!--
    If this pull request fully addresses an issue, use:
    Fixes #xxxx

    If this pull request partially addresses an issue, use:
    Part of #xxxx

    If this pull request addresses multiple issues, put them in multiple lines:
    Fixes #xxxx
    Fixes #yyyy

    Format the title of the pull request with most relevant issue number as follows:
    [#xxxx] Title
-->
Fixes #2016 

### Proposed commit message
<!--
    Propose a detailed commit message for this pull request within the triple backticks below.
    Wrap lines at 72 characters.

    Guide on how to write a good commit message:
    https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message
-->
```
Fix bug where some params were not retained after refresh

The issue pertains to breakdown of file types disappearing after a
refresh, but it seems to apply to all params that were last in the URL
due to symbols not being filtered out correctly.

Let's fix this issue to make sure the state of the report stays the
same when refreshing.
```

### Other information
<!--
    Are there other relevant information, such as special testing instructions,
    which will help the reviewer better understand the code?

    You may also include a brief description of why the problem occurred.
-->
The bug seems to have been caused by the URL containing `#/` at the end, which would cause the last param to not be parsed correctly. I couldn't definitively find what caused the anchor to appear in the URL, so I filtered it out of the URL when decoding `hashParams`. Let me know if this behaviour should be changed!

Regarding the Cypress tests, #2017 seems to be outdated/some of the tests don't seem to be checking for the correct thing, at least on my machine. If I should investigate this further in anticipation of its being merged, let me know as well!

sopa301 commented:
Googling a bit suggests that this appending of `#/` is caused by Vue's [Hash Mode](https://router.vuejs.org/guide/essentials/history-mode), which appends a `#` followed by the route `/` (the default one) at the end of the URL. This could be a regression from the implementation of Vue Router.

I tested this using the 2324S1 CS2103T RepoSense dashboard. Some evidence for this hypothesis:

- Using `#/widget` gives a different view of the webpage.
- Using `#/test` gives a blank webpage

That said, the current fix is good enough to make it work. I propose 2 possible paths in the future:

1. Rewrite the code to ensure Vue Router's internal routing is finished and the URL is fixed before it is processed by `api.ts` (not recommended due to complexity)
2. Refactor `api.ts` to capture this consideration more distinctly.

jonasongg commented:
@sopa301 Thanks for the suggestions! 

Just wanted to link these related issues here since it seems they are caused by this same problem: #2058 #1987

sopa301 reviewed (approved):
LGTM. As commented above, this is sufficient to fix the bug, but may warrant further refactoring.

ckcherry23 reviewed (commented):
Thanks for your inputs @jonasongg and @sopa301. I think this temporary fix is a reasonable solution for now as it resolves the current issue without introducing new issues. We can switch to the Vue Router HTML5 mode in the future. 

@jonasongg Since this fix is somewhat unconventional, can you add a comment in the code explaining the rationale behind the change? This will help future developers understand the context and reasoning behind this decision.

Regarding the Cypress tests, we should ideally be adding a test case for the bug fixed in this PR itself. Let us wait to see the progress of PR #2017 over the next few days and make a call over the changes needed.

MarcusTXK reviewed (commented):
Nice fix. Agree with @ckcherry23 that a comment to explain the logic for future developers would be useful.

Regarding the tests, I think perhaps @jonasongg could reference the code written in #2017 and add the updated tests, then co-author credit could be added to @nseah21.

jonasongg commented:
@ckcherry23 @MarcusTXK Thanks for the reviews! Just to confirm, I should add the `codeView_renderFilterHash.cy.js` file from #2017 here and add tests that test this new expected behaviour?

ckcherry23 commented:
@jonasongg #2017 was just merged. You can look into what additions/changes are required on the tests now, thanks!

MarcusTXK commented:
@jonasongg After discussion, we decided to merge in #2017, so you can work on the merged test directly on your branch.

jonasongg commented:
Just added the comment and some test cases! For checked file types, I added some tests to make sure it was testing the last file type to appear in the URL (yml) as that was what was being affected by the bug. Also added some non exhaustive test cases for the code panel options including one for authorshipFileGlob that is most likely the cause for #2058 and #1987. Let me know if I should add more/change anything!

ckcherry23 commented on frontend/cypress/tests/codeView/codeView_renderFilterHash.cy.js:
```diff
@@ -309,31 +309,184 @@ describe('render filter hash', () => {
     cy.url()
       .should('not.contain', 'gradle');
 
+    cy.url()
+      .should('not.contain', 'yml');
+
```
Can we make it clear why gradle and yml were the 2 file types selected here with a comment?
For example,
```
// Assumption: gradle is the first file type and yml is the last file type

ckcherry23 commented on frontend/cypress/tests/codeView/codeView_renderFilterHash.cy.js:
```diff
@@ -309,31 +309,184 @@ describe('render filter hash', () => {
+      .should('contain', 'reverseAuthorshipOrder=true');
+  });
+
+  it('code panel: filter by glob: url params should persist after change and reload', () => {
```
This function only tests after one reload, but there is a bug in the filter by glob string after 2 reloads. 
Hence, just wanted to add a note that we must add a test with 2 page reloads when the bug in #1987 gets fixed.

ckcherry23 reviewed (approved):
LGTM! Just have a few minor comments

jonasongg commented:
Should be clarified now!

MarcusTXK reviewed (approved):
LGTM, I think you did a great job
